### ISSUE TITLE
Improve deadlock detection

### ISSUE THREAD

**Issue by vasild** [bitcoin/bitcoin#19157]
Background
---
If two or more threads acquire mutexes in different order, that could cause a deadlock. Currently we have two mechanisms for detecting that - our [DEBUG_LOCKORDER](https://github.com/bitcoin/bitcoin/blob/bdedfcf/src/sync.cpp#L134) and the [thread sanitizer](https://clang.llvm.org/docs/ThreadSanitizer.html).

Problem
---
Both methods may fail to detect some deadlocks:
* `DEBUG_LOCKORDER` only detects deadlocks that involve two mutexes: `A => B, B => A`. It will not detect `A => B, B => C, C => A`.
* Both methods require all relevant code paths to execute in the same invocation of the program. `A => B, restart the program, B => A` will not be detected. I guess this is how the bug which https://github.com/bitcoin/bitcoin/pull/19132 fixes sneaked in.
* The thread sanitizer fails to detect a deadlock from this test:
<details>
<summary>[1] test case causing undetected deadlock by the thread sanitizer</summary>

```cpp
BOOST_AUTO_TEST_CASE(deadlock_3)
{
#if 0
    // detected by DEBUG_LOCKORDER
    // not detected by the thread sanitizer (when DEBUG_LOCKORDER is disabled)
    constexpr size_t n_threads = 2;
#else
    // deadlock is not detected by either one
    constexpr size_t n_threads = 3;
#endif

    // thread0 locked mutex0
    // thread1 locked mutex1
    // thread2 locked mutex2
    // thread0 trying to lock mutex1
    // thread1 trying to lock mutex2
    // thread2 trying to lock mutex0 <-- deadlocks

    class Event
    {
    public:
        void signal()
        {
            std::unique_lock<std::mutex> lk(m_mutex);
            m_occurred = true;
            m_cond.notify_all();
        }
        void wait()
        {
            std::unique_lock<std::mutex> lk(m_mutex);
            m_cond.wait(lk, [&]() { return m_occurred; });
        }

    private:
        bool m_occurred{false};
        std::condition_variable m_cond;
        std::mutex m_mutex;
    };

    std::mutex printf_mutex;
    auto printf_sync = [&](const char* format, ...) {
        va_list ap;
        va_start(ap, format);

        {
            std::unique_lock<std::mutex> lk(printf_mutex);
            vprintf(format, ap);
        }

        va_end(ap);
    };

    std::array<Mutex, n_threads> mutexes;
    std::array<std::thread, n_threads> threads;
    std::array<Event, n_threads> locked_own;
    std::array<Event, n_threads> try_to_lock_next;

    auto f = [&](size_t i) {
        LOCK(mutexes[i]);
        printf_sync("thread%zu locked mutex%zu\n", i, i);
        locked_own[i].signal();

        try_to_lock_next[i].wait();

        const size_t next = (i + 1) % n_threads;
        printf_sync("thread%zu trying to lock mutex%zu\n", i, next);
        LOCK(mutexes[next]);
    };

    for (size_t i = 0; i < n_threads; ++i) {
        threads[i] = std::thread{f, i};
    }

    for (size_t i = 0; i < n_threads; ++i) {
        locked_own[i].wait();
    }

    for (size_t i = 0; i < n_threads; ++i) {
        try_to_lock_next[i].signal();
    }

    for (size_t i = 0; i < n_threads; ++i) {
        threads[i].join();
    }
}
```
</details>

Solution
---
Attach a predefined integer to each mutex that denotes the locking order in relation to other mutexes. So, whenever attempting to lock a mutex a thread would check if the previous mutex it acquired has a lower number.

This would require some thorough considerations when adding a new mutex - when is it going to be used, what other mutexes are going to be held at the time the new one is acquired and what other mutexes may be acquired while holding the new one. That is a good practice and should be done anyway. Such a mechanism will enforce it.

**Comment by MarcoFalke** [comment]
> Both methods may fail to detect some deadlocks

It seems they fail to detect a deadlock, but isn't the deadlock happening good enough of an indication that something is wrong? For example running the tests and observing a reproducible timeout should be enough to hint to a deadlock, no?

**Comment by vasild** [comment]
Well, the `DEBUG_LOCKORDER` checks and the deadlock detection from tsan exist because deadlocks happening are not good enough to analyze the problem. A sporadic timeout on CI once per week, or worse, on the users' computers without any printouts is very difficult to analyze.

Deadlocks are often elusive and not 100% repoducable. Even in the test case above most of the code is to ensure that the deadlock actually happens :)

**Comment by MarcoFalke** [comment]
Concept ACK

**Comment by vasild** [comment]
I submitted the testcase at https://github.com/google/sanitizers/issues/1258. It looks like a bug to me.

**Comment by sipa** [comment]
@vasild Are you familiar with the dining philosopher's problem?

**Comment by vasild** [comment]
@sipa, in the context of the dining philosophers problem I propose the [resource hierarchy solution](https://en.wikipedia.org/wiki/Dining_philosophers_problem#Resource_hierarchy_solution) where all resources (mutexes) are related by order.

**Comment by sipa** [comment]
@vasild I'm not sure that works in the presence of multiple partially overlapping sets of dependent locks.

Let's say the following lock orders exist: A->B A->C B->C (which is not a deadlock). Assume A->C is grabbed first, so you assign A=1, C=2; then A->B is grabbed, assigning B=3. Now B->C is grabbed which is considered a reversal and incorrectly treated as a deadlock.

**Comment by vasild** [comment]
Sorry, I should have elaborated more - by "predefined" I mean hardcoded in the source code (not assigned dynamically at runtime).

In the above case the developer(s) would assign A=1, B=2, C=3. Then, if the code has a bug and locks in different order, e.g. first B, then A, just executing that code path is enough to catch the problem. No need to have two contradicting executions A->B & B->A like now.

I counted about 70 mutexes in the entire code base, including tests. If it turns out that this is too big to do in one go we can start with defining the order between some mutexes and assign "wildcard" to the rest, skipping the hierarchy check for them.

**Comment by MarcoFalke** [comment]
This sounds similar to https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#acquired-before-acquired-after, but I heard someone say they are not actually implemented in libcxx <sup>[Citation needed]</sup>

**Comment by hebasto** [comment]
> This sounds similar to https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#acquired-before-acquired-after, but I heard someone say they are not actually implemented in libcxx [Citation needed]

https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#acquired-before-and-acquired-after-are-currently-unimplemented

**Comment by sipa** [comment]
@vasild I think that's a very ugly solution, as it requires global coordination over the whole codebase. It makes components less reusable, and means that the introduction of code that now grabs two previously unrelated locks may result in requiring a renumbering of several locks.

As performance does not matter that much, I think you could just run a naive cycle-finding algorithm (DFS traversal, caching which nodes you've already hit) to find issues.

**Comment by hebasto** [comment]
FWIW, a custom `hierarchical_mutex` mutex type and its usage are provided by Anthony Williams in his _C++ Concurrency in Action_, SE, 2019 (3.2.4 Deadlock: the problem and a solution -- Use a lock hierarchy).

**Comment by vasild** [comment]
All, thanks for the input! Next I will see if I can turn this into some code, so that we have something concrete to discuss.

PS The cycle-finding algorithm is what TSan already does.

**Comment by adamjonas** [comment]
@vasild did you ever investigate this further?

**Comment by vasild** [comment]
No, but I don't think much investigation is needed. I just need to sit down and implement a hierarchical mutex (the [resource hierarchy solution](https://en.wikipedia.org/wiki/Dining_philosophers_problem#Resource_hierarchy_solution)) and see how it will be accepted by the community. I have worked with a [simple version of that in the InnoDB storage engine in MySQL](https://github.com/mysql/mysql-server/blob/fbdaa4def30d269bc4de5b85de61de34b11c0afc/storage/innobase/include/sync0types.h#L59) that was written in the 90s (thanks, Heikki! :heart:). I think it can be done in a more elegant way with the new language extensions that we have now in 2022.

**Comment by willcl-ark** [comment]
This issue hasn’t attracted much interest from other contributors in quite some time.

Given that, it doesn’t seem important enough to keep open indefinitely. I’m going to close it for now due to lack of activity, but pull requests or renewed discussion are always welcome.

Comment here if you think this should be re-opened.
